### ISSUE TITLE
Newline for each item in JSON array

### ISSUE THREAD

**Issue by Robinfr** [prettier/prettier#2716]
Is there an option to place each item in a JSON array on a new line regardless of the line of the length? I have quite a lot of the following:

```json
[
    ["thing1"],
    ["thing2"]
]
```

And would like keep them this way instead of the whole thing becoming one line. I also think in general placing each item in an array on a new line is preferable.

**Comment by azz** [comment]
> I also think in general placing each item in an array on a new line is preferable. 

That's a little bit tricky.

```json
[
    ["thing1", "thing2"],
    ["thing3"]
]
```

would have to be formatted as

```json
[
    [
        "thing1",
        "thing2"
    ],
    ["thing3"]
]
```

**Comment by Robinfr** [comment]
Yes I see what you mean. Personally I would always go for the second one, but I understand why someone would want the first one. 

Is there any way this can be configurable?

**Comment by nickiaconis** [comment]
+1 to the second format @azz posted, for JS too. When paired (in JS) with trailing commas, that format leads to clean version control diffs.

I would recommend a similar format for JS object literals as well (which is why I'm looking through Prettier issues in the first place).

**Comment by lydell** [comment]
An idea worth discussing: What if we took the “3-or-more” rule from chained method calls and applied it to arrays (both in JS and in JSON)? One or two items would always be single-line (if it fits), while three or more items would always be multiline (even if it fits in a single line).

```js
let a = [];
let b = [1];
let c = [1, 2];
let d = [
  1,
  2,
  3,
];
```

**Comment by Robinfr** [comment]
Wouldn't it be better if there was an option to choose after how many items the split happens?

**Comment by maraisr** [comment]
What about if its based on column count - ie > 80 then new line them in some sort of snake? Just thinking like, single char arrays would take up way too many lines.. Or is it more about what makes the diff prettier (no pun intended :sunglasses:)?

```diff
- [1, 2, 3, 4, 5, 6]
+ [1, 2, 3
+ , 4, 5]
```
> :snake: version.

**Comment by Robinfr** [comment]
@maraisr that's also fine in my opinion.

**Comment by Robinfr** [comment]
I'm not sure more discussion is possible here @azz ..

**Comment by bennyn** [comment]
Regarding new lines for array items... 

I do prefer:

```javascript
queryCategories() {
  return Category.query().select([
    'id',
    'color',
    'name',
  ]);
}
```

Over:

```javascript
queryCategories() {
  return Category.query().select(['id', 'color', 'name']);
}
```

**Reasons:** 

1. Although the first example is longer, it makes it easier for me to add new array items, because they are listed one per each line which makes it easy to insert (or remove) an item by cutting (`Ctrl + X`) or copy & pasting (`Ctrl + C` & `Ctrl + V`) lines. 
1. There are line sorters (like [Lines Sorter](https://plugins.jetbrains.com/plugin/5919-lines-sorter) & [Sort line tools](http://plugins.netbeans.org/plugin/45925/sort-line-tools)) which work by selecting the lines, that you want to have sorted, with your mouse. When taking the Prettier approach (all arguments within one line), you cannot select multiple lines because there is just one. 
1. Another reason why I favor the first option is because Prettier would also go that way (if the statement with the array exceed `printWidth`). Having sometimes array items one per each line (statement > `printWidth`) and sometimes not (statement < `printWidth`) doesn't look very consistent (and leads to longer Git diffs).

Let me give you another example...

**Example A: Without Prettier**

```javascript
exports.down = (knex, Promise) => {
  return Promise.all([
    knex.schema.alterTable('categories', (table) => table.dropColumns('created_at', 'updated_at')),
  ]);
};
```

**Example B: With Prettier**

```javascript
exports.down = (knex, Promise) => {
  return Promise.all([knex.schema.alterTable('categories', table => table.dropColumns('created_at', 'updated_at'))]);
};
```

It's easier to add another `knex.schema.alterTable` statement to **Example A**, don't you think so?

**Comment by bennyn** [comment]
Regarding new lines for array items... 

I do prefer:

```javascript
queryCategories() {
  return Category.query().select([
    'id',
    'color',
    'name',
  ]);
}
```

Over:

```javascript
queryCategories() {
  return Category.query().select(['id', 'color', 'name']);
}
```

**Reasons:** 

1. Although the first example is longer, it makes it easier for me to add new array items, because they are listed one per each line which makes it easy to insert (or remove) an item by cutting (`Ctrl + X`) or copy & pasting (`Ctrl + C` & `Ctrl + V`) lines. 
1. There are line sorting plugins (like [Lines Sorter](https://plugins.jetbrains.com/plugin/5919-lines-sorter) & [Sort line tools](http://plugins.netbeans.org/plugin/45925/sort-line-tools)) which work by selecting the lines, that you want to have sorted, with your mouse. When taking the Prettier approach (all items within one line), you cannot select multiple lines because there is just one. 
1. Another reason why I favor the first option is because Prettier would also go that way (if the statement with the array exceed `printWidth`). Having sometimes array items one per each line (statement length > `printWidth`) and sometimes not (statement length < `printWidth`) doesn't look very consistent to me and leads to bigger Git diffs.

Let me give you another example...

**Example A: Without Prettier**

```javascript
exports.down = (knex, Promise) => {
  return Promise.all([
    knex.schema.alterTable('categories', (table) => table.dropColumns('created_at', 'updated_at')),
  ]);
};
```

**Example B: With Prettier**

```javascript
exports.down = (knex, Promise) => {
  return Promise.all([knex.schema.alterTable('categories', table => table.dropColumns('created_at', 'updated_at'))]);
};
```

It's easier to add another `knex.schema.alterTable` statement to **Example A**, don't you think so?

**Comment by suchipi** [comment]
What settings are you using prettier with that output example B? Default settings output:
```js
exports.down = (knex, Promise) => {
  return Promise.all([
    knex.schema.alterTable("categories", table =>
      table.dropColumns("created_at", "updated_at")
    )
  ]);
};
```
But also, for what it's worth, I would flag a Promise.all with 1 item in it in a code review anyway; I would write something like this until there were two items:
```js
exports.down = (knex, Promise) =>
  knex.schema.alterTable("categories", table =>
    table.dropColumns("created_at", "updated_at")
  );
```
That's arguably a matter of preference, though (optimizing for diff vs. optimizing for now).

**Comment by bennyn** [comment]
@suchipi: These are my Prettier settings:

```javascript
{
  "bracketSpacing": false,
  "jsxBracketSameLine": false,
  "printWidth": 120,
  "requirePragma": false,
  "semi": true,
  "singleQuote": true,
  "tabWidth": 2,
  "trailingComma": "es5",
  "useTabs": false
}
```

I guess my example looks different than yours because I use a `printWidth` of 120 (instead of the recommended 80).

Regarding the `Promise.all`... There I like to quote Lea Verou who once wrote: "Always strive to make code that can be modified with the least number of edits".

**Comment by jeznag** [comment]
I have the opposite desire. I'd like to maximise the number of array items on one line. I have test data that looks like this. I'd much prefer to have the values array over 1-2 lines instead of 48. Any way to achieve that?

```
const timeIntervalData = [
  {
    timePeriod: '日',
    values: [
      1,
      1,
      1,
      1,
      1,
      2,
      2,
      2,
      2,
      2,
      2,
      2,
      2,
      2,
      2,
      2,
      2,
      0,
      0,
      0,
      0,
      2,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
    dayIndex: '0',
  },
```

**Comment by JounQin** [comment]
A configurable option would be great. Prettier output `.json` is really not pretty like other json formatter tools.

**Comment by suchipi** [comment]
It's unfortunate that prettier-ignore comments usually can't be used in JSON, or I'd recommend that for @jeznag's situation...

**Comment by duailibe** [comment]
I don't think @jeznag's example is JSON though... he can use `// prettier-ignore` there.

**Comment by suchipi** [comment]
Ah, you're totally right

**Comment by mmermerkaya** [comment]
[Clang-Format](https://github.com/angular/clang-format) handles this beautifully by using single-line formatting if you don't add a trailing comma, and multi-line formatting if you do.

```
let x = {a, b, c};
```

```
let x = {
  a,
  b,
  c,
};
```

Unfortunately Prettier enforces a single rule for trailing commas which makes this impossible.

**Comment by nikolaycviv** [comment]
Any further work on this? I would like to have different options available. Like @jeznag suggested, I have such array in .js file...

**Comment by j-f1** [comment]
@nikolaycviv Right now, your only option is to put a `// prettier-ignore` comment before the array.

**Comment by psy901** [comment]
@j-f1 Thanks for the tip! 
I was having hard time watching my array being formatted like 
```js
values: [
      1,
      1,
      1,
      1,
      1,
      2,
      2,
      2,
      2,
      2,
      2,
      2,
];
```

now it looks better with `// prettier-ignore`

**Comment by phamhongphuc** [comment]
I'm changing my default formatter from **jsbeautify** to **prettier** because I need a typescript formatter but I got a problem.

With [js-beautify](https://github.com/beautify-web/js-beautify):

```javascript
console.log({
    a: 1,
    b: 2,
    c: 3,
}, {
    d: 4,
    e: 5
});
```

With **prettier**:

```javascript
console.log(
    {
        a: 1,
        b: 2,
        c: 3
    },
    {
        d: 4,
        e: 5
    }
);
```
Is there an option to resolve this?

**Comment by dvlsg** [comment]
Just to chime in with another use case, the array format rules often make `Map` harder to read in the code I've been writing.

For example, the following:
```
const map = new Map([
  [Constants.A, functionA],
  [Constants.B, functionB],
  [Constants.C, functionC],
]);
```

Gets turned into this (assuming the line max length is large enough):
```
const map = new Map([[Constants.A, functionA], [Constants.B, functionB], [Constants.C, functionC]]);
```

Using `prettier-ignore` works fine, of course, but I do find myself tossing `prettier-ignore` around pretty frequently when instantiating maps.

**Comment by j-f1** [comment]
@dvlsg That’s why we recommend using a `printWidth` of 80 🙂

**Comment by dvlsg** [comment]
Haha, that's fair. I use 80 on any new projects I start. and I definitely prefer it. Unfortunately, the one I'm currently working on (with a team) is using 120.

Just ran across this issue while checking to see if there was a way I could hint to `prettier` to keep array values on separate lines without losing the rest of `prettier`'s formatting capabilities for that chunk of code -- similar to what `clang-format` does with trailing commas -- and figured I'd toss my use case on the pile.

**Comment by JoshMcCullough** [comment]
Print width is not the answer here, IMO, @j-f1. I want a wide print width so that simple function calls are not needlessly wrapped. But I also want Prettier to respect my newlines if I am trying to make an array (or a function call) more readable. 

Seems like there are two options here, other than using just print width.

1. Prettier could respect what the user has already done. E.g. if an array spans more than one line, put each array item on its own line; if an array is all on a single line, leave it on a single line (but prettify the spacing); if that single line exceeds the print width, then put each element on its own line as it does now.

2. Prettier could determine the complexity of the array. If any element is complex (e.g. a function call, plain object, long string, etc. then it would put each element on its own line; if all array items are primitives e.g. `[1, 2, 3, ..., 1000]`, they would stay on a single line and wrap once the print width is reached; short vs. long strings could be handled separately -- short: all elements on single line; long: all elements on own line.

**Comment by suchipi** [comment]
@dvlsg @MariuzM I opened an issue here: https://github.com/prettier/prettier/issues/5780

**Comment by chrisblossom** [comment]
Can we just treat arrays the same way we do as plain objects? If the developer adds a blank line, respect it for the rest of the array. I've never understood why this was accepted for plain objects, but not for arrays. This would also solve #5780 #6000 (probably others too).

I typically add an unwanted empty comment to accomplish this...

```js
const arr = [
	//
	1,
	2,
]
```

**Comment by JoshMcCullough** [comment]
>I typically add an unwanted empty comment to accomplish this...

And destroy the integrity of my beautiful code? Never!

Seriously though, if you have to add comments to appease your tools, something is wrong with the tools (or your expectations of the tools, as well be it's too believe in this case).

**Comment by j-f1** [comment]
Maybe we should pass an array literal passed to `new Map(...)` as an object literal, since that’s what it functions as. However, that may not be a great idea since moving the array out of the `new Map()` call would cause it to be formatted differently ¯\_(ツ)_/¯

**Comment by waldyrious** [comment]
> Can we just treat arrays the same way we do as plain objects? If the developer adds a blank line, respect it for the rest of the array. I've never understood why this was accepted for plain objects, but not for arrays. 

Agreed. If anything, this behavior would make Prettier _more_ consistent, not less. The same [rationale for why multi-line objects are preserved](https://prettier.io/docs/en/rationale.html#multi-line-objects) applies exactly for the array case. Having a different behavior for each case is what goes against spirit of the project, IMHO.

**Comment by noltron000** [comment]
Just want to chime in here and agree to the point that having a few configurable options would be highly preferable:
1. force all arrays multi-line
2. keep all arrays single-line
3. split with column print-width (default)

Right now users can _only_ use option 3, which actually reduces the prettiness of code in many places. The git diff argument is absolutely a perfect argument for this, as is the ugliness of the `//` empty-comment work-around, and honestly the `// prettier-ignore`  comment solution. The churn is real!

Just my 2 cents :)

**Comment by lydell** [comment]
This has been improved a little by #6694

**Comment by j-f1** [comment]
I’d like to note something: GitHub reports that there 4.2 million matches for [`eslint-disable`](https://github.com/search?q=eslint-disable&type=Code) and 1.3 million matches for [`eslint-disable-next-line`](https://github.com/search?q=eslint-disable-next-line&type=Code), while there are [only](https://github.com/search?q=prettier-ignore&type=Code) 37k matches for `prettier-ignore` on all of public GitHub. Even accounting for usage differences (if Prettier had as many downloads as ESLint, there’d be about 55k `prettier-ignore`s), people seem to be on the whole very willing to accept Prettier’s style.

**Comment by JounQin** [comment]
@j-f1 That's no true. Eslint report warning or error while prettier just change it without complaints.

**Comment by waynebloss** [comment]
@j-f1 You could also conclude that people are not willing to clutter up their code with _even more_ comment codes just to fight against a formatter that's mildly infuriating them on a regular basis by constantly wrapping and unwrapping things.

Plus, `eslint-ignore` is a lot older and a lot more well known than `prettier-ignore`.

**Comment by benwiley4000** [comment]
Not going to solve it for everyone, but I was able to get this to work by setting my `--print-width` to 1 - it will always wrap when it gets to a new entry!

**Comment by LakMoore** [comment]
Can we hear from the maintainers of their opinion on why Prettier handles JSON Arrays differently to JSON objects?  I think Prettier users are generally agreed that we like the way Objects are formatted and we are confused as to why Arrays are handled differently.  This definitely does not need to be configurable.  The proposal looks like more consistent behaviour to me, but perhaps I am missing something?

Having read this thread, every use case would be satisfied by making this change.  Nested arrays in Maps would work because the dev would provide a newline after the first Open bracket and none after every other.

If someone made a PR to make this change would it be accepted?

**Comment by wmakley** [comment]
Just came here due to this. Why are arrays not treated the same as objects? This rule is actually reducing the readability and increasing the git diffs in my code base, two things that are contrary to the prettier goals I would have thought.

**Comment by guilhermetod** [comment]
That's really the only thing keeping me from using prettier. I would use it through my entire project if it wasn't for this rule, but because of it I can only see myself using it on html files.

At least let us disable it and use ESLint rules.

**Comment by Baehn** [comment]
In Jest the function "toMatchInlineSnapshot" uses prettier to format inline snapshots which are json-strings. My test files become unreadable because of the many newlines inserted for objects and arrays. It would be really great to be able to modify this behaviour!

**Comment by lydell** [comment]
@Baehn You might also be interested in https://github.com/prettier/prettier/issues/5807

**Comment by varthc** [comment]
Any news/update on this? We were starting to implement Prettier in all the Angular projects we have, but this is definitely a huge problem and we had to put it on hold.
All the alternatives to solve this are hacks and the issue and workarounds impact in the readability of the code. 
How hard is to be consistent and have the same rule(s) that Prettier has for multi-line objects? Serious question.

**Comment by SimonFarrugia** [comment]
I desperately need this. I have code that represents DOM structure, and need it to be readable (HTML like).

**Comment by fisker** [comment]
Is "json-stringfy" parser you are looking for?

**Comment by SimonFarrugia** [comment]
@fisker No I don't think so. What I have is  something similar to react-hyperscript.

**Comment by rb-mwindh** [comment]
It's the same for me. I'm just trying to integrate prettier into our workflow when it turned my angular imports into oneliners. Since this messes up my git diffs, for now prettier is a no-go for me. That' s a shame.

**Comment by AndreasA** [comment]
To be honest if it is just readability I would not care but if I have arrays with fix values that are edited by multiple persons, 
it is extremely tricky to resolve a merge conflict if everything is in one line. 
Whereas if it is multiple lines, most IDEs provide an easy way to resolve such conflicts and there might not even be conflicts but the single line version will nearly always have such issues.

Probably not necessary if the array has just one value though. and maybe some other scenarios too but in general the multi line approach seems safer regarding git merges.

**Comment by IT-CASADO** [comment]
For me this is also the most (if not the only) annoying issue with prettier.

And the reason to reduce merge conflicts should be enough to change this formatting rule.

Either by hard coded rule or by configuration.

**Comment by hypeJunction** [comment]
This formatting is driving me crazy. You have to be able to use `git blame` to see who changed what array values.

**Comment by OJFord** [comment]
I found this issue wanting the same for objects, e.g.:
```js
import {
  One,
  Two,
} from "whatever";
```
becomes:
```js
import { One, Two } from "whatever";
```
(with default config).

Perhaps one solution is to take after [black](//github.com/psf/black) (python formatter) and consider the manual trailing comma as a signal to keep one-per-line?

Also, another workaround - I found that adding an inline comment (any comment, doesn't have to be an ignore) inside the array/object stops it:
```js
import {
  //
  One,
  Two,
} from "whatever";
```

**Comment by shiomax** [comment]
Also ran into this. Im working on a project that can read DPM Data Matrix Codes. So, I have test data that´s for instance a 14x14 2d array of booleans. It makes a ton more sence to arrange this as a grid instead of newlines between every element, so you can actually read it more or less. Im probably never gonna change the test data Ill just add more. But now my 14x14 arrays take up 225 lines instead of 16 and it´s virtually impossible to read as a human.

**Comment by TheNightmareX** [comment]
Agree. It makes me really uncomfortable to see my arrays like `imports` in Angular be formatted into  a single line!!

```ts
@NgModule({
  imports: [
    BrowserModule,
    AppRoutingModule,
    BrowserAnimationsModule,
  ],
})
export class AppModule {}
```

```ts
@NgModule({
  imports: [BrowserModule, AppRoutingModule, BrowserAnimationsModule],
})
export class AppModule {}
```

**Comment by taletski** [comment]
+1 to configuring multi-line

Looks like a lot of folks are frustrated by of the absence of this feature: prettier/prettier-vscode#352

**Comment by Gregorein** [comment]
+1 for ability to configure this.

**Comment by Gregorein** [comment]
being able to configure for arrays in general would be nice, but I don't think we'll ever get that: https://prettier.io/docs/en/option-philosophy.html

**Comment by IT-CASADO** [comment]
In general I'm a big fan of the "no option philosophy". 

But the current implementation of this array rule is difficult to read/resolve code in merge conflicts.

If an additional option is not an option, then prettier should think about to change the current implementation (in the next major/minor release).

**Comment by julix-unity** [comment]
What @IT-CASADO said. It's fine that it's opinionated, but this opinion is wrong based on Prettier's goals (i.e. less merge conflicts, etc.)

**Comment by varthc** [comment]
@electrovir thanks very much for the plugin! I'll try it when I come back from vacations.
We forked prettier to solve this issue, but I think your plugin if it works in our projects, it's the right way to go. Thanks again!

**Comment by lppedd** [comment]
The mentioned plugin seems stuck trying to adapt to Prettier v3.  
Since Prettier maintainers do not seem to like it in the main repo, it would be nice if we could get it (again) as a plugin, so I think it would useful for the plugin author to get some "official" help.

**Comment by electrovir** [comment]
FYI (@lppedd) [`prettier-plugin-multiline-arrays `](https://www.npmjs.com/package/prettier-plugin-multiline-arrays) now supports Prettier v3
